### ISSUE TITLE
Use unicode Paths on Windows

### ISSUE THREAD

**Issue by abergmeier-dsfishlabs** [bazelbuild/bazel#2101]
### Description of the problem / feature request / question:

Bazel on Windows currently passes paths to Windows using postfix A names (`C:/whatever`). We found that it is REALLY easy to get to a point where you exceed the `MAX_PATH` limit. So you should always use Unicode paths on Windows using the prefix `\\?\`.
And please change it now because afterwards I feel it will be too hard to change anymore.

### If possible, provide a minimal example to reproduce the problem:

Install Visual Studio to:

    C:/Program Files/1234567890_and_life_is_great/and_I_am_creating_a_really_horrible_path_which_is_actually_not_that_uncommon_on/yes_it_is_horrible_but_hey_a_bit_of_horrible_is_always_fun_right_right_yessss_sure_it_is_just_perfect_and_should_be_really_great_so_do_it_like_this.windows

and set `BAZEL_VS` to this path.

### Environment info
* Operating System: Windows 8.1

* Bazel version (output of `bazel info release`): 0.4.0

###  Have you found anything relevant by searching the web? (e.g. GitHub issues, email threads in the [bazel-discuss@googlegroups.com](https://groups.google.com/forum/#!forum/bazel-discuss) archive)

There are tons of discussions online for when and how you should use this.

### Anything else, information or logs or outputs that would be helpful?
Not sure whether Java does this under the hood already. Then mostly only the direct interfacing with Windows (tools) needs to be fixed.

**Comment by dslomov** [comment]
Yes, thanks, we are aware of this. Unfortunately, the problem is kind of endemic: at least on Windows 7 cmd.exe does not work with \\?\ paths well. I also think cl.exe had problems with those paths, esp in versions before 2015. 

Are you running into issues without artificially installing VS into crazy locations?

**Comment by abergmeier-dsfishlabs** [comment]
> Yes, thanks, we are aware of this.

Great.

> Unfortunately, the problem is kind of endemic: at least on Windows 7 cmd.exe does not work with \?\ paths well. I also think cl.exe had problems with those paths, esp in versions before 2015.

Another reason to switch from `cmd.exe` to PowerShell as much as possible. Perhaps there also needs to be global `--experimental_unicode_paths` switch to select whether one wants to use them on Windows?

> Are you running into issues without artificially installing VS into crazy locations?

The VS was just an example.
We do have these problems in our own built tool and _really_ long source file paths. Had to do some Junctions in the past to shorten the path lengths.

Chances are good that we might encounter these problems when using Bazel to built these files.

**Comment by Helcaraxan** [comment]
This is also becoming an issue on my side with @petemounce. We don't have any crazy paths (i.e the path to our `WORKSPACE` is only 2 folders deep and has 23 chars (including the `WORKSPACE` dir's name).

If we want to use this more widely for builds on Windows we can not expect developers to systematically make junctions like `C:\A\<workspace-dir>` to make it work.

TL;DR +1 on the issue.

**Comment by petemounce** [comment]
I'm on Windows 10 and suffer this.

**Comment by petemounce** [comment]
Related to #2145?

**Comment by laszlocsomor** [comment]
@petemounce : Yes it is related, though not the same. https://github.com/bazelbuild/bazel/issues/2181 is the same issue as this one. Closing this in favor of that, since I've been submitting fixes referencing that bug.
